File: synara/src/app/features/matrix-ipc/streamBody.ts
/**
 * R0.3 residual / REV-005: topic-typed snapshot/delta body validation.
 *
 * Bodies remain JSON objects on the wire. This module binds each stream topic
 * to owned DTO containers (via matrix-dto parsers) and rejects secret-like
 * fields and media-like numeric byte arrays at the envelope boundary.
 */

import {
  FORBIDDEN_WIRE_FIELD_NAMES,
  hasForbiddenWireFields,
  isObject,
  parseNotificationCandidate,
  parseReceipt,
  parseRoomMember,
  parseRoomSummary,
  parseSecurityStatus,
  parseTimelineItem,
  parseTypingSnapshot,
} from '../matrix-dto';
import type { StreamTopic } from './stream';

function isForbiddenFieldName(name: string): boolean {
  return (FORBIDDEN_WIRE_FIELD_NAMES as readonly string[]).includes(name);
}

/** Recursively reject forbidden keys and pure numeric arrays (media bytes). */
export function rejectForbiddenAndMediaBytes(value: unknown): boolean {
  if (Array.isArray(value)) {
    if (value.length > 0 && value.every((v) => typeof v === 'number')) {
      return false;
    }
    return value.every((child) => rejectForbiddenAndMediaBytes(child));
  }
  if (isObject(value)) {
    if (hasForbiddenWireFields(value)) return false;
    for (const [key, child] of Object.entries(value)) {
      if (isForbiddenFieldName(key)) return false;
      if (!rejectForbiddenAndMediaBytes(child)) return false;
    }
    return true;
  }
  return true;
}

function onlyKnownKeys(o: Record<string, unknown>, allowed: readonly string[]): boolean {
  return Object.keys(o).every((k) => allowed.includes(k));
}

function parseObjectArray<T>(value: unknown, parseOne: (v: unknown) => T | null): T[] | null {
  if (!Array.isArray(value)) return null;
  const out: T[] = [];
  for (const item of value) {
    const parsed = parseOne(item);
    if (!parsed) return null;
    out.push(parsed);
  }
  return out;
}

function parseAccountDataEvent(
  value: unknown,
): { type: string; content: Record<string, unknown> } | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  if (!onlyKnownKeys(value, ['type', 'content'])) return null;
  if (typeof value.type !== 'string') return null;
  const content = value.content === undefined ? {} : isObject(value.content) ? value.content : null;
  if (content === null) return null;
  if (!rejectForbiddenAndMediaBytes(content)) return null;
  return { type: value.type, content };
}

function parsePresenceEntry(value: unknown): {
  userId: string;
  presence: string;
  statusMsg?: string;
  lastActiveTs?: number;
} | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  if (!onlyKnownKeys(value, ['userId', 'presence', 'statusMsg', 'lastActiveTs'])) return null;
  if (typeof value.userId !== 'string' || typeof value.presence !== 'string') return null;
  if (value.statusMsg !== undefined && typeof value.statusMsg !== 'string') return null;
  if (
    value.lastActiveTs !== undefined &&
    (typeof value.lastActiveTs !== 'number' || !Number.isFinite(value.lastActiveTs))
  ) {
    return null;
  }
  return {
    userId: value.userId,
    presence: value.presence,
    statusMsg: typeof value.statusMsg === 'string' ? value.statusMsg : undefined,
    lastActiveTs: typeof value.lastActiveTs === 'number' ? value.lastActiveTs : undefined,
  };
}

function parseSendQueueItem(value: unknown): {
  localId: string;
  roomId: string;
  state: string;
} | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  if (!onlyKnownKeys(value, ['localId', 'roomId', 'state'])) return null;
  if (
    typeof value.localId !== 'string' ||
    typeof value.roomId !== 'string' ||
    typeof value.state !== 'string'
  ) {
    return null;
  }
  return { localId: value.localId, roomId: value.roomId, state: value.state };
}

/**
 * Validate snapshot/delta body for a stream topic.
 * Returns true when the body is a typed, privacy-safe projection for the topic.
 */
export function validateStreamTopicBody(topic: StreamTopic, body: unknown): boolean {
  if (!isObject(body)) return false;
  if (!rejectForbiddenAndMediaBytes(body)) return false;

  switch (topic) {
    case 'room_list': {
      if (!onlyKnownKeys(body, ['rooms'])) return false;
      if (body.rooms === undefined) return true;
      return parseObjectArray(body.rooms, parseRoomSummary) !== null;
    }
    case 'timeline': {
      if (!onlyKnownKeys(body, ['items'])) return false;
      if (body.items === undefined) return true;
      return parseObjectArray(body.items, parseTimelineItem) !== null;
    }
    case 'members': {
      if (!onlyKnownKeys(body, ['members'])) return false;
      if (body.members === undefined) return true;
      return parseObjectArray(body.members, parseRoomMember) !== null;
    }
    case 'typing': {
      if (!onlyKnownKeys(body, ['rooms'])) return false;
      if (body.rooms === undefined) return true;
      return parseObjectArray(body.rooms, parseTypingSnapshot) !== null;
    }
    case 'receipts': {
      if (!onlyKnownKeys(body, ['receipts'])) return false;
      if (body.receipts === undefined) return true;
      return parseObjectArray(body.receipts, parseReceipt) !== null;
    }
    case 'account_data': {
      if (!onlyKnownKeys(body, ['events'])) return false;
      if (body.events === undefined) return true;
      return parseObjectArray(body.events, parseAccountDataEvent) !== null;
    }
    case 'presence': {
      if (!onlyKnownKeys(body, ['entries'])) return false;
      if (body.entries === undefined) return true;
      return parseObjectArray(body.entries, parsePresenceEntry) !== null;
    }
    case 'notification_candidates': {
      if (!onlyKnownKeys(body, ['candidates'])) return false;
      if (body.candidates === undefined) return true;
      return parseObjectArray(body.candidates, parseNotificationCandidate) !== null;
    }
    case 'crypto_status': {
      if (!onlyKnownKeys(body, ['status'])) return false;
      if (body.status === undefined) return true;
      return parseSecurityStatus(body.status) !== null;
    }
    case 'send_queue': {
      if (!onlyKnownKeys(body, ['items'])) return false;
      if (body.items === undefined) return true;
      return parseObjectArray(body.items, parseSendQueueItem) !== null;
    }
    default: {
      // Exhaustiveness: unknown topics already rejected by isStreamTopic.
      const _exhaustive: never = topic;
      return _exhaustive;
    }
  }
}

File: crates/synara-core/src/transport/stream_body.rs
//! R0.3 residual: bind snapshot/delta bodies to topic-typed DTO containers.
//!
//! REV-005 requires every supported stream topic to carry bounded, typed bodies
//! and to reject secret-like fields and media byte arrays at the envelope
//! boundary. Bodies remain JSON objects on the wire; validation deserializes
//! them into topic-specific owned shapes using domain DTOs (P1.4).

use serde::Deserialize;
use serde_json::Value;

use crate::dto::{
    NotificationCandidate, Receipt, RoomMember, RoomSummary, SecurityStatus, TimelineItem,
    TypingSnapshot, FORBIDDEN_WIRE_FIELD_NAMES,
};

use super::stream::StreamTopic;

/// Room list snapshot/delta body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RoomListStreamBody {
    #[serde(default)]
    pub rooms: Vec<RoomSummary>,
}

/// Timeline snapshot/delta body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TimelineStreamBody {
    #[serde(default)]
    pub items: Vec<TimelineItem>,
}

/// Members snapshot/delta body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MembersStreamBody {
    #[serde(default)]
    pub members: Vec<RoomMember>,
}

/// Typing snapshot/delta body (multi-room projection).
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TypingStreamBody {
    #[serde(default)]
    pub rooms: Vec<TypingSnapshot>,
}

/// Receipts snapshot/delta body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ReceiptsStreamBody {
    #[serde(default)]
    pub receipts: Vec<Receipt>,
}

/// Account-data stream body (typed shell; content objects still scanned).
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AccountDataStreamBody {
    #[serde(default)]
    pub events: Vec<AccountDataEventWire>,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AccountDataEventWire {
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default)]
    pub content: serde_json::Map<String, Value>,
}

/// Presence stream body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PresenceStreamBody {
    #[serde(default)]
    pub entries: Vec<PresenceEntryWire>,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PresenceEntryWire {
    pub user_id: String,
    pub presence: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status_msg: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_active_ts: Option<u64>,
}

/// Notification candidates snapshot/delta body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct NotificationCandidatesStreamBody {
    #[serde(default)]
    pub candidates: Vec<NotificationCandidate>,
}

/// Crypto / security status stream body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CryptoStatusStreamBody {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<SecurityStatus>,
}

/// Send-queue / local-echo status stream body.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SendQueueStreamBody {
    #[serde(default)]
    pub items: Vec<SendQueueItemWire>,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SendQueueItemWire {
    pub local_id: String,
    pub room_id: String,
    /// Product local-echo state wire name (`sending`, `sent`, `failed`, …).
    pub state: String,
}

/// Validate a snapshot/delta `body` for the given stream topic.
///
/// Rules (REV-005 residual):
/// 1. Body must be a JSON object.
/// 2. Forbidden secret / media field names are rejected anywhere in the tree.
/// 3. Pure numeric arrays (media-like byte payloads) are rejected.
/// 4. Body deserializes into the topic-owned container (`deny_unknown_fields`).
pub fn validate_stream_topic_body(
    topic: StreamTopic,
    body: &Value,
) -> Result<(), serde_json::Error> {
    if !body.is_object() {
        return Err(serde::de::Error::custom(
            "stream body must be a JSON object",
        ));
    }
    reject_forbidden_and_media_bytes(body)?;
    match topic {
        StreamTopic::RoomList => {
            let _: RoomListStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::Timeline => {
            let _: TimelineStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::Members => {
            let _: MembersStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::Typing => {
            let _: TypingStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::Receipts => {
            let _: ReceiptsStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::AccountData => {
            let _: AccountDataStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::Presence => {
            let _: PresenceStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::NotificationCandidates => {
            let _: NotificationCandidatesStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::CryptoStatus => {
            let _: CryptoStatusStreamBody = serde_json::from_value(body.clone())?;
        }
        StreamTopic::SendQueue => {
            let _: SendQueueStreamBody = serde_json::from_value(body.clone())?;
        }
    }
    Ok(())
}

fn reject_forbidden_and_media_bytes(value: &Value) -> Result<(), serde_json::Error> {
    match value {
        Value::Object(map) => {
            for (key, child) in map {
                if is_forbidden_field_name(key) {
                    return Err(serde::de::Error::custom(format!(
                        "forbidden wire field in stream body: {key}"
                    )));
                }
                reject_forbidden_and_media_bytes(child)?;
            }
            Ok(())
        }
        Value::Array(items) => {
            if !items.is_empty() && items.iter().all(|v| v.is_number()) {
                return Err(serde::de::Error::custom(
                    "media-like numeric byte array forbidden in stream body JSON IPC",
                ));
            }
            for child in items {
                reject_forbidden_and_media_bytes(child)?;
            }
            Ok(())
        }
        _ => Ok(()),
    }
}

fn is_forbidden_field_name(name: &str) -> bool {
    FORBIDDEN_WIRE_FIELD_NAMES.contains(&name)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn empty_object_valid_for_all_topics() {
        let empty = json!({});
        for topic in StreamTopic::ALL {
            validate_stream_topic_body(*topic, &empty).unwrap_or_else(|e| {
                panic!("empty body should be valid for {:?}: {e}", topic);
            });
        }
    }

    #[test]
    fn room_list_accepts_summary_fixture_shape() {
        let body = json!({
            "rooms": [{
                "roomId": "!room:example.org",
                "name": "General",
                "membership": "join",
                "isDirect": false,
                "isEncrypted": true,
                "unreadCount": 0,
                "highlightCount": 0,
                "markedUnread": false
            }]
        });
        validate_stream_topic_body(StreamTopic::RoomList, &body).unwrap();
    }

    #[test]
    fn room_list_rejects_unknown_keys() {
        let body = json!({ "rooms": [], "extra": true });
        assert!(validate_stream_topic_body(StreamTopic::RoomList, &body).is_err());
    }

    #[test]
    fn rejects_secret_field_nested() {
        let body = json!({
            "rooms": [{
                "roomId": "!r:e.org",
                "membership": "join",
                "isDirect": false,
                "isEncrypted": false,
                "unreadCount": 0,
                "highlightCount": 0,
                "markedUnread": false,
                "accessToken": "s3cret"
            }]
        });
        let err = validate_stream_topic_body(StreamTopic::RoomList, &body).unwrap_err();
        assert!(
            err.to_string().contains("forbidden") || err.to_string().contains("accessToken"),
            "err={err}"
        );
    }

    #[test]
    fn rejects_media_byte_array() {
        let body = json!({ "items": [1, 2, 3, 4] });
        // timeline expects objects in items; numeric array fails either media rule or item parse
        assert!(validate_stream_topic_body(StreamTopic::Timeline, &body).is_err());
        let nested = json!({ "rooms": [], "blob": [0, 255, 128] });
        // unknown key rooms body — still fail media if structure were open; room_list deny_unknown
        assert!(validate_stream_topic_body(StreamTopic::RoomList, &nested).is_err());
        let media_in_account = json!({
            "events": [{
                "type": "m.media",
                "content": { "mediaBytes": [1, 2, 3] }
            }]
        });
        assert!(validate_stream_topic_body(StreamTopic::AccountData, &media_in_account).is_err());
    }

    #[test]
    fn timeline_accepts_empty_items() {
        validate_stream_topic_body(StreamTopic::Timeline, &json!({ "items": [] })).unwrap();
    }

    #[test]
    fn wrong_topic_shape_rejected() {
        // timeline-shaped body on room_list topic
        let body = json!({ "items": [] });
        assert!(validate_stream_topic_body(StreamTopic::RoomList, &body).is_err());
    }
}
